### PR TITLE
 issue-1499

### PR DESCRIPTION
**Pull request by jarosik10**
for: https://github.com/openstax/unified/issues/1499

**Testing:**
Go to `config.books.json` and delete two books: `13ac107a-f15f-49d2-97e8-60ab2e3b519c`, `9d8df601-4f12-4ac1-8224-b450bf739e5f`.

`13ac107a-f15f-49d2-97e8-60ab2e3b519c` has a backup in `release.development.json`.

`13ac107a-f15f-49d2-97e8-60ab2e3b519c` stands for Algebra and Trigonometry `../books/algebra-and-trigonometry/pages/1-introduction-to-prerequisites`
`9d8df601-4f12-4ac1-8224-b450bf739e5f` stands for American Government 2e `../books/american-government-2e/pages/1-introduction`

### CODE REVIEW

**Comment by jarosik10** [comment]
@TomWoodward do we want to add the book config gateway as a `AppService`?

**Comment by TomWoodward** [comment]
yea you'll need to configure the base url and pass it through like the other gateways

**Comment by jarosik10** [comment]
From where can I retrieve the base url for `.../rex/release.json`?

**Comment by TomWoodward** [comment]
you can load it relatively in the browser, but for example it can be found here https://openstax.org/rex/release.json

**Comment by jarosik10** [comment]
@TomWoodward 
How can it be tested? because `rex/release.json` can't be accessed on `localhost` or `herokuapp`.

**Comment by TomWoodward** [comment]
you should set it up so that if it fails to load release.json it fails silently and doesn't cause an error to be displayed

to test that it actually works you can add a test file similar to the way that environment.development.json works

**Comment by jarosik10** [comment]
@TomWoodward could you generally review the logic? And do we want to completely remove importing BOOKS config directly and replace it with the gateway?

File: src/app/content/hooks/locationChange/resolveContent.ts
Comment (TomWoodward) [inline src/app/content/hooks/locationChange/resolveContent.ts:181]: i think the reference has a version on it, you might not have to do this here

File: src/gateways/createBookConfigLoader.ts
Comment (TomWoodward) [inline src/gateways/createBookConfigLoader.ts:21]: there is an `acceptStatus` helper that you could use here

File: src/gateways/createBookConfigLoader.ts
Comment (TomWoodward) [inline src/gateways/createBookConfigLoader.ts:27]: it would be nice to have this put a warning in sentry, but definitely not pop an error to the user

**Review by TomWoodward — CHANGES_REQUESTED** [review]
you may want to actually base your changes on piotr's rap refactor because it simplifies the types in resolveContent i think

File: src/app/content/hooks/locationChange/resolveContent.ts
Comment (jarosik10) [reply src/app/content/hooks/locationChange/resolveContent.ts:159]: oh yah. It should accept undefined here.

File: src/test/mocks/bookConfigLoader.ts
Comment (PiotrKozlowski) [inline src/test/mocks/bookConfigLoader.ts:8]: Probably it should be mocked in the similar way to `mocks/archiveLoader.ts`

File: src/index.tsx
Comment (PiotrKozlowski) [inline src/index.tsx:55]: this should be configured like other urls

File: src/app/content/hooks/locationChange/resolveContent.ts
Comment (PiotrKozlowski) [reply src/app/content/hooks/locationChange/resolveContent.ts]: this is added in https://github.com/openstax/rex-web/pull/1133

File: src/gateways/createBookConfigLoader.ts
Comment (PiotrKozlowski) [inline src/gateways/createBookConfigLoader.ts:33]: does it need to be here? it seems to me that it can be moved directly to the `loadRemoteBookConfig`

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]
I didn't look at `resolveContent.ts ` because it will change after merging https://github.com/openstax/rex-web/pull/1133

File: src/app/content/hooks/locationChange/resolveContent.ts
Comment (jarosik10) [reply src/app/content/hooks/locationChange/resolveContent.ts]: Should I wait for #1133 to be merged? @PiotrKozlowski @TomWoodward

File: src/index.tsx
Comment (jarosik10) [reply src/index.tsx:55]: What exactly do you mean? via config?

Comment (PiotrKozlowski) [reply src/index.tsx:55]: you can do it similar to `archiveUrl`, `buyPrintConfigUrl` etc. from this file

File: src/app/content/hooks/locationChange/resolveContent.ts
Comment (PiotrKozlowski) [reply src/app/content/hooks/locationChange/resolveContent.ts]: You could either merge my branch into yours and change the base for this PR or you could just wait and resolve conflicts when my PR is merged - idk when this will happen

File: src/index.tsx
Comment (jarosik10) [reply src/index.tsx:55]: @PiotrKozlowski 
I wonder if the path to `release.json` file should be placed in config or just move all the logic for creating the `fetchURL` to the `bookConfigLoader` gateway?

Comment (PiotrKozlowski) [reply src/index.tsx:55]: I would stick to the same format as for other gateways, so you pass a `baseUrl` as you currently do, and `/rex/release.json` is added in `createBookConfigLoader` as it is now
the only change is to replace `window.location.origin` with a const `bookConfigUrl` that will be created with `assertDefined` in the same way like for example `buyPrintConfigUrl`

Comment (jarosik10) [reply src/index.tsx:55]: Okay, but there is question if we would like to make the `BOOK_CONFIG_URL` from `config.js` to receive the url as:
- `BOOK_CONFIG_URL: window.location.origin`
- or `BOOK_CONFIG_URL: typeof(window) === 'undefined' ? 'https://openstax.org' : window.location.origin`

And the naming seems weird. If this property only contains the origin path, maybe it should be named the way showing it. To make it feel reusable.

However it still feels awkward form me to store it in config, but maybe I'm missing something :D

Comment (jarosik10) [reply src/index.tsx:55]: @TomWoodward what do you think?

File: src/gateways/createBookConfigLoader.ts
Comment (PiotrKozlowski) [reply src/gateways/createBookConfigLoader.ts:33]: this is still not resolved

File: src/gateways/createBookConfigLoader.ts
Comment (PiotrKozlowski) [inline src/gateways/createBookConfigLoader.ts:34]: I think it would be good to create `const cachedBooks = {...BOOKS}` and use this variable to store there versions loaded from the remote book configs so we don't request 2 times the same book from the remote config

File: src/config.js
Comment (PiotrKozlowski) [inline src/config.js:27]: this would be probably `process.env.ORIGIN_URL || 'https://openstax.org'`
and in the dev config it could be `typeof(window) === 'undefined' ? 'https://openstax.org' : window.location.origin,` or 
`typeof(window) === 'undefined' ? 'https://openstax.org' : '',`

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]
Should this new gateway be used in places like `getUrlParamsForBook`, `getCanonicalUrlParams` functios?

**Comment by jarosik10** [comment]
I've find out that when there is a missing book in local book config (look testing in pr's description) the `href`s of links in ToC change from:
`https://localhost:3000/books/algebra-and-trigonometry/pages/1-1-real-numbers-algebra-essentials`
to:
`https://localhost:3000/books/13ac107a-f15f-49d2-97e8-60ab2e3b519c@32.2/pages/1-1-real-numbers-algebra-essentials`

**Comment by TomWoodward** [comment]
@jarosik10 the reverse routing checks to see if the book is at the configured version and uses that to decide if it renders the regular link or the versioned link. did you not update that to use your new config gateway?

**Comment by jarosik10** [comment]
@TomWoodward 
I'm working on it. But it will require a lot of code changing. Changin _getUrlParamsForBook_ and _getBookPageUrlAndParams_ in `urlUtils.ts` will affect a lot of components and _domVisitor_, _urlChecker_ scripts.

**Comment by TomWoodward** [comment]
any scripts that aren't running the app are probably fine to continue accessing the config json directly. anything in the app that would also be running in the browser needs to use the gateway. the logic in the url param helpers that decides if it needs the versions in the url, that should use the gateway. unfortunately that will mean making a lot of url logic async and i'm unsure how gnarly that is going to be

File: src/gateways/createBookConfigLoader.ts
Comment (PiotrKozlowski) [inline src/gateways/createBookConfigLoader.ts:35]: my understanding was that we should have 2 methods - one synchronous and the second one asynchronous 
both of them should access `cachedBooks` variable
the async one will be used in `resolveContent` and it will update `cachedBooks` if it wasn't found previously
the sync one will just return version from `cachedBooks`

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]

File: src/gateways/createBookConfigLoader.ts
Comment (jarosik10) [reply src/gateways/createBookConfigLoader.ts:35]: To be consistent I will have to replace all `BOOKS[id]` calls in the project with the sync function from the `bookConfigLoader`, right? But in the other hand it will make some components need to receive or load `AppServices['BookConfigLoader']`, which will require some refactoring (like in this [PR](https://github.com/openstax/rex-web/pull/1153), but without making some functions async).

Comment (PiotrKozlowski) [reply src/gateways/createBookConfigLoader.ts:35]: I think this will be okay. The other option would be to just create a module that exports these functions instead of creating it as a gateway and do not pass it to the services

Comment (PiotrKozlowski) [reply src/gateways/createBookConfigLoader.ts:35]: @TomWoodward Do you have an opinion on this?

Comment (TomWoodward) [reply src/gateways/createBookConfigLoader.ts:35]: it might be ok for the synchronous function to be a regular export that references a `let books` at the top of the module

File: src/gateways/createBookConfigLoader.ts
Comment (PiotrKozlowski) [inline src/gateways/createBookConfigLoader.ts:48]: I would call this function differently and change the type of the first one to always return a promise

File: src/app/content/utils/canonicalUrl.ts
Comment (PiotrKozlowski) [inline src/app/content/utils/canonicalUrl.ts:27]: `assertDefined` or map through `cannonicals` and add `version` to them

File: src/app/content/hooks/locationChange.spec.ts
Comment (PiotrKozlowski) [inline src/app/content/hooks/locationChange.spec.ts:208]: I think you could just `mockImplementation` on `getBookVersionFromUUID`

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]

File: src/app/content/hooks/locationChange.spec.ts
Comment (jarosik10) [reply src/app/content/hooks/locationChange.spec.ts:208]: Is it really needed? It seems easier to mock the config there.

File: src/app/content/utils/canonicalUrl.ts
Comment (PiotrKozlowski) [inline src/app/content/utils/canonicalUrl.ts:29]:
```suggestion
      `We've already filtered out books that are not in the `BOOK` configuration`
```

**Review by PiotrKozlowski — APPROVED** [review]

File: src/app/content/hooks/locationChange.spec.ts
Comment (TomWoodward) [reply src/app/content/hooks/locationChange.spec.ts:208]: i agree that it would be more straight forward to use a spy on `getBookVersionFromUUID`, but its fine like this

File: src/config.d.ts
Comment (TomWoodward) [inline src/config.d.ts:13]: origin url for what? is this used for loading the rex configs? oh i see below you're handling undefined window, is this only for loading rex configs when pre-rendering? because when not pre-rendering you could just not put a host and use a relative url like `/rex/release.json`

when pre-rendering our local config will definitely be newer than the remote, we should not go out of our way to make fetching the remote work. in the gateway we could skip the remote check if window is undefined

File: src/gateways/createBookConfigLoader.ts
Comment (TomWoodward) [inline src/gateways/createBookConfigLoader.ts:29]: i can't imagine this being anything but a network failure, which will just be noise in sentry, worth reporting it but maybe make it a warning

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: src/config.d.ts
Comment (TomWoodward) [reply src/config.d.ts:13]: if you're not using the gateway in the scripts, then you don't need the skip check in the gateway, but i don't understand why you need the origin url config, you can just reference with a relative url like `/rex/release.json`

Comment (PiotrKozlowski) [reply src/config.d.ts:13]: I think I've suggested adding it to the config since I saw a bunch of this type of variables there, but indeed relative url seems better

**Comment by TomWoodward** [comment]
@jarosik10 it doesn't look like you've updated this since my review

**Comment by jarosik10** [comment]
@TomWoodward Sorry. Changes are pushed.

File: src/config.js
Comment (TomWoodward) [inline src/config.js:27]: this isn't necessary right?

File: src/gateways/createBookConfigLoader.ts
Comment (TomWoodward) [inline src/gateways/createBookConfigLoader.ts:19]: this is the type that the books object already has, why do you need to redeclare `BookConfig` and `BookVersion` here?

File: src/gateways/createBookConfigLoader.ts
Comment (TomWoodward) [inline src/gateways/createBookConfigLoader.ts:42]: why do you have all this logic in here? why don't you just assign books over cachedBooks and return `getBookVersionFromUUIDSync`?

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: src/gateways/createBookConfigLoader.ts
Comment (jarosik10) [reply src/gateways/createBookConfigLoader.ts:19]: yea, actually it is not needed to point it here `const cachedBooks = { ...BOOKS } as BookConfig;`, but these interfaces are showing what data types the function returns.

File: src/gateways/createBookConfigLoader.ts
Comment (jarosik10) [reply src/gateways/createBookConfigLoader.ts:42]: this is what you mean?
```js
getBookVersionFromUUID: (uuid: string): Promise<BookVersion | undefined> => {
      return cachedBooks[uuid] ? Promise.resolve(cachedBooks[uuid]) : loadRemoteBookConfig().then((books) => {
        if (books) {
          cachedBooks = books;
        }
        return getBookVersionFromUUIDSync(uuid);
      });
    },
});
```

Comment (TomWoodward) [reply src/gateways/createBookConfigLoader.ts:42]: i think so, that works right?

File: src/gateways/createBookConfigLoader.ts
Comment (TomWoodward) [reply src/gateways/createBookConfigLoader.ts:19]: in the case of the function you can just use `typeof BOOKS` right?

**Review by TomWoodward — CHANGES_REQUESTED** [review]
there are still some outstanding comments

File: src/gateways/createBookConfigLoader.ts
Comment (TomWoodward) [inline src/gateways/createBookConfigLoader.ts:5]: i don't think this needs to be exported

File: src/setupProxy.js
Comment (TomWoodward) [inline src/setupProxy.js:210]: this json file will become old and weird very quickly, can you construct a json response here based on the actual books config? 

or maybe since the json config will never really be used in development, since it is only ever used when the book check is missed, maybe leave the release.development.json but leave the books section empty, that way its there if someone wants to test something and add a book but it doesn't have content that will need to be updated to continue making sense

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: src/gateways/createBookConfigLoader.ts
Comment (jarosik10) [reply src/gateways/createBookConfigLoader.ts:5]: It is imported in `src/test/mocks/bookConfigLoader.ts`.

Comment (TomWoodward) [reply src/gateways/createBookConfigLoader.ts:5]: you'd be better off setting the mock `getBookVersionFromUUID` to `ReturnType<typeof createBookConfigLoader>['getBookVersionFromUUID']` so that its forced to match the actual type instead of deriving the same type again

**Review by TomWoodward — CHANGES_REQUESTED** [review]

**Review by TomWoodward — APPROVED** [review]

**Pull request merged**
